Interval: add default implementations for most methods

maarzt opened the pull request:
Most methods in the interface `Interval` can have a default implementation. 
These default implementations will prevent a lot of bugs and code duplication.
Implementing an Interval becomes as easy as overloading three methods:
```java
public class SimpleInterval implements Interval
{
	private final long[] min;
	private final long[] max;

	public SimpleInterval( long[] min, long[] max )
	{
		assert min.length == max.length;
		this.min = min;
		this.max = max;
	}

	@Override
	public long min( int d )
	{
		return min[ d ];
	}

	@Override
	public long max( int d )
	{
		return max[ d ];
	}

	@Override
	public int numDimensions()
	{
		return min.length;
	}
}
```

maarzt commented:
I checked if adding this default implementations has any influence on performance.
Performance stays the same. 

The code I used for benchmarking: https://github.com/maarzt/imglib2/compare/master...benchmark-interval-default-implementations

```
################ Interval without changes ###############

Benchmark                                             Mode  Cnt        Score        Error  Units
DefaultIntervalBenchmark.abc                         thrpt    8   774081.514 ±  41143.894  ops/s
DefaultIntervalBenchmark.benchmarkMin                thrpt    8  2046413.208 ± 102335.106  ops/s
DefaultIntervalBenchmark.benchmarkRealMin            thrpt    8  2331398.003 ±  42055.958  ops/s
DefaultIntervalBenchmark.intervalContains            thrpt    8    52633.326 ±   2050.379  ops/s
DefaultIntervalBenchmark.intervalContainsRealPoints  thrpt    8  3888439.603 ± 461708.769  ops/s
DefaultIntervalBenchmark.iterateExtended             thrpt    8     9756.587 ±    274.826  ops/s
DefaultIntervalBenchmark.splitIntervall              thrpt    8     7148.196 ±   3052.784  ops/s

################ Interval with default implementations #############

Benchmark                                             Mode  Cnt        Score        Error  Units
DefaultIntervalBenchmark.abc                         thrpt    8   755103.282 ±  33699.039  ops/s
DefaultIntervalBenchmark.benchmarkMin                thrpt    8  2177057.950 ±  51290.428  ops/s
DefaultIntervalBenchmark.benchmarkRealMin            thrpt    8  2239097.764 ± 153554.875  ops/s
DefaultIntervalBenchmark.intervalContains            thrpt    8    53153.321 ±   1523.556  ops/s
DefaultIntervalBenchmark.intervalContainsRealPoints  thrpt    8  3931321.971 ± 212744.809  ops/s
DefaultIntervalBenchmark.iterateExtended             thrpt    8     9727.041 ±    391.607  ops/s
DefaultIntervalBenchmark.splitIntervall              thrpt    8     7222.925 ±   3215.695  ops/s
```

ctrueden commented:
Awesome, I'm in big favor of migrating default implementations from abstract layers into interfaces wherever possible. Right now, implementing key ImgLib2 interfaces requires a lot of methods; and since ImgLib2 uses composition so much, you can't always extend a single abstract class to avoid it. Default methods for the win! :trophy:

tpietzsch commented:
Yes, I also think it's time for a bit of cleaning up, using java 8 features.

However, the `RealInterval` methods should be implemented as default methods in `RealInterval` to avoid duplication

ctrueden commented:
"Every default method should be pushed upstream as far as possible, but no farther" 😜

tpietzsch commented:
@maarzt Could you move the `RealInterval` default methods to `RealInterval`?
(And run the benchmark again to see whether it makes a difference to have the more complicated  hierarchy with default methods inherited from different parents.)

maarzt commented:
As requested I moved some default method implementations from `Interval` to `RealInterval` and `Dimensions`. I also added default method implementations to `Localizable` and `RealLocalizable`.
And I did the benchmark again.
```
**Before**
Benchmark                                             Mode  Cnt        Score        Error  Units
DefaultIntervalBenchmark.benchmarkMin                thrpt    8  2149053.201 ± 185376.861  ops/s
DefaultIntervalBenchmark.benchmarkRealMin            thrpt    8  2296187.559 ± 106860.403  ops/s
DefaultIntervalBenchmark.intervalContains            thrpt    8    50827.843 ±   1405.206  ops/s
DefaultIntervalBenchmark.intervalContainsRealPoints  thrpt    8  3585043.399 ± 695232.202  ops/s
DefaultIntervalBenchmark.iterateExtended             thrpt    8     3269.245 ±    136.902  ops/s
DefaultIntervalBenchmark.splitIntervall              thrpt    8     6929.482 ±   2940.917  ops/s

**After (with default method implementation**
Benchmark                                             Mode  Cnt        Score        Error  Units
DefaultIntervalBenchmark.benchmarkMin                thrpt    8  2150682.270 ± 141313.709  ops/s
DefaultIntervalBenchmark.benchmarkRealMin            thrpt    8  2314714.556 ±  82895.721  ops/s
DefaultIntervalBenchmark.intervalContains            thrpt    8    50241.894 ±   5093.540  ops/s
DefaultIntervalBenchmark.intervalContainsRealPoints  thrpt    8  3811206.108 ± 346780.226  ops/s
DefaultIntervalBenchmark.iterateExtended             thrpt    8     9459.974 ±    673.724  ops/s
DefaultIntervalBenchmark.splitIntervall              thrpt    8     6992.915 ±   2890.582  ops/s
```

Performance stays the same.
@tpietzsch I think it can be merged?
